### PR TITLE
WIP: Fix Navigator focus requests when offstage

### LINKED ISSUE #45634: Offstage Navigators steal focus from an onstage Navigator
On initial launch of our app, a route on an offstage navigator requests focus. This causes focus traversal to happen on a widget tree that's offscreen, which is very confusing.

Minimal reproduction:

```dart
import 'package:flutter/widgets.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  testWidgets("offstage nav focus", (tester) async {
    Widget _childNavRouteBuilder(BuildContext context, _, __) {
      return Text('foobar');
    }

    Route<void> _childNavRoute(RouteSettings settings) {
      return PageRouteBuilder(
        pageBuilder: _childNavRouteBuilder,
      );
    }

    final topLevelNav = GlobalKey<NavigatorState>();
    final navA = GlobalKey<NavigatorState>();
    final navB = GlobalKey<NavigatorState>();
    final navC = GlobalKey<NavigatorState>();

    await tester.pumpWidget(Directionality(
      textDirection: TextDirection.ltr,
      child: Navigator(
        key: topLevelNav,
        onGenerateRoute: (settings) {
          return PageRouteBuilder(
            pageBuilder: (context, _, __) {
              return Stack(
                children: <Widget>[
                  Offstage(
                    child: Navigator(key: navA, onGenerateRoute: _childNavRoute),
                  ),
                  Offstage(
                    offstage: false,
                    child: Navigator(key: navB, onGenerateRoute: _childNavRoute),
                  ),
                  Offstage(
                    child: Navigator(key: navC, onGenerateRoute: _childNavRoute),
                  ),
                ],
              );
            },
          );
        },
      ),
    ));

    await tester.pumpAndSettle();

    expect(topLevelNav.currentState.focusScopeNode.hasFocus, isTrue);
    expect(navA.currentState.focusScopeNode.hasFocus, isFalse);
    expect(navB.currentState.focusScopeNode.hasFocus, isTrue); // Actual: false
    expect(navC.currentState.focusScopeNode.hasFocus, isFalse); // Actual: true
  });
}
```

cc @gspencergoog

### PR DESCRIPTION
**Pull request by gspencergoog**
## Description

This fixes a bug where if you had navigators that were offstage, they would still try and request the focus, which is incorrect behavior.

## Related Issues
- Fixes https://github.com/flutter/flutter/issues/45634
## Tests
- Added a test for offstage nested navigators.
## Breaking Change

- [X] No, this is *not* a breaking change.

### CODE REVIEW

**Comment by gspencergoog** [comment]
cc @ds84182

File: packages/flutter/lib/src/widgets/navigator.dart [hunk 1]
```diff
@@ -136,7 +136,10 @@ abstract class Route<T> {
   @mustCallSuper
   TickerFuture didPush() {
     return TickerFuture.complete()..then<void>((void _) {
-      navigator.focusScopeNode.requestFocus();
+      final Offstage offstageWidget = navigator.context.findAncestorWidgetOfExactType<Offstage>();
+      if (offstageWidget == null || !offstageWidget.offstage) {
```
Comment (ds84182) [inline packages/flutter/lib/src/widgets/navigator.dart]: Would this have correct behavior if there are multiple Offstage widgets present in the tree, with some onstage and some not?

Comment (goderbauer) [reply packages/flutter/lib/src/widgets/navigator.dart]: If you have a Navigator wrapped in an Offstage with offstage == false and everything wrapped in an Offstage with offstage == true this probably wouldn't do the right thing.

**Comment by goderbauer** [comment]
There's a somewhat related problem with Offstage and Heros: https://github.com/flutter/flutter/pull/44890

**Comment by gspencergoog** [comment]
Actually, after talking this over with @Hixie, I'm going to recommend that instead of walking the tree to find things (which is O(n) bad), that the focus behavior be set explicitly.  I've added an `autofocus` setting to `RouteSettings` so that the focus behavior can be controlled.

The PR has changed significantly, so PTAL.

**Comment by goderbauer** [comment]
I wonder if this is actually a viable approach: Very often you don't have access to the RouteSettings of a Route to set this flag (e.g. when you use MaterialApp.routes or MaterialApp.home). 

Would it instead be possible to wrap something in some kind of a FocusScope configured so that nothing in this subtree can get the focus? Similarly to e.g. how TickerMode turns off animations in a subtree?

**Comment by goderbauer** [comment]
(PR triage): Is this one superseded by https://github.com/flutter/flutter/pull/46168 and can be closed?

**Comment by gspencergoog** [comment]
Yes.  Closing.

**Pull request closed without merging**